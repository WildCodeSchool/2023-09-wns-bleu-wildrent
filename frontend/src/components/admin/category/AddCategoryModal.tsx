import React from 'react';
import FormInput from '@/components/FormInput';
import { useAddCategoryMutation } from '@/graphql/generated/schema';
import client from '@/graphql/client';
import { useAlert } from '@/components/providers/AlertContext';

const fields = [
  {
    label: 'Nom de la catÃ©gorie',
    id: 'name',
    type: 'text',
    placeholder: 'Entrez le nom de la catÃ©gorie',
    required: true,
  },
  {
    label: 'Description de la catÃ©gorie',
    id: 'description',
    type: 'textarea',
    placeholder: 'Entrez la description',
    required: true,
  },
  {
    label: 'Image de la catÃ©gorie',
    id: 'image',
    type: 'text',
    placeholder: "URL de l'image",
    required: true,
  },
];

function AddCategoryModal({
  isOpen,
  onClose,
  onCategoryAdded,
}: {
  isOpen: boolean;
  onClose: () => void;
  onCategoryAdded: (category: any) => void;
}) {
  if (!isOpen) return null;
  const [createCategory, { loading }] = useAddCategoryMutation();
  const { showAlert } = useAlert();
  const handleSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    if (!e.currentTarget.checkValidity()) return;
    const formData = new FormData(e.currentTarget);
    const formJSON = Object.fromEntries(formData.entries());

    try {
      const response = await createCategory({
        variables: {
          name: formJSON.name.toString(),
          image: formJSON.image.toString(),
          description: formJSON.description.toString(),
        },
      });
      if (response.data?.addCategory) {
        console.log('ðŸš€ ~ handleSubmit ~ response.data?.addCategory:', response.data?.addCategory);
        showAlert('success', 'Category added successfully', 3000);
        onCategoryAdded(response.data.addCategory);
        onClose();
      } else {
        showAlert('error', 'Error adding category', 3000);
      }
    } catch (error) {
      showAlert('error', 'Network or query error while adding category', 3000);
      console.error('Error adding category', error);
    } finally {
      client.resetStore();
    }
  };

  return (
    <div>
      <input
        type="checkbox"
        id="category_modal"
        className="modal-toggle"
        checked={isOpen}
        readOnly
      />
      <div className="modal" role="dialog">
        <div className="modal-box">
          <h3 className="text-lg font-bold">Ajout d'une nouvelle catÃ©gorie</h3>
          <form className="flex flex-col gap-4 p-4 border rounded" onSubmit={handleSubmit}>
            {fields.map((field) => (
              <FormInput
                key={field.id}
                id={field.id}
                label={field.label}
                placeholder={field.placeholder}
                inputType={field.type}
                required={field.required}
              />
            ))}
            <button type="submit" className="btn btn-primary" disabled={loading}>
              Add
            </button>
          </form>
        </div>
        <label className="modal-backdrop" htmlFor="category_modal" onClick={onClose}>
          Close
        </label>
      </div>
    </div>
  );
}

export default AddCategoryModal;
