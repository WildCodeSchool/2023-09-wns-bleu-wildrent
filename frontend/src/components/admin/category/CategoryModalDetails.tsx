import React from 'react';

export type CategoryModalProps = {
  isOpen: boolean;
  onClose: () => void;
  category?: {
    name: string;
    description: string;
    image: string;
  };
};

function CategoryDetailsModal({ isOpen, onClose, category }: CategoryModalProps) {
  if (!isOpen) return null;

  return (
    <div>
      <input
        type="checkbox"
        id="category_modal"
        className="modal-toggle"
        checked={isOpen}
        readOnly
      />
      <div className="modal" role="dialog">
        <div className="modal-box">
          <h3 className="text-lg font-bold">{category?.name}</h3>
          <p className="py-4">{category?.description}</p>
          <img src={category?.image} alt="Image de la catÃ©gorie" />
        </div>
        <label className="modal-backdrop" htmlFor="category_modal" onClick={onClose}>
          Fermer
        </label>
      </div>
    </div>
  );
}

export default CategoryDetailsModal;
