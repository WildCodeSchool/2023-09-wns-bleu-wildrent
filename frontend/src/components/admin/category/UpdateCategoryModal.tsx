import React from 'react';
import FormInput from '@/components/FormInput';
import { useUpdateCategoryMutation, AllCategoriesAdminQuery } from '@/graphql/generated/schema';
import client from '@/graphql/client';

type Category = AllCategoriesAdminQuery['allCategories'][0];

const fields = [
  {
    label: 'Nom de la catégorie',
    id: 'name',
    type: 'text',
    placeholder: 'Entrez le nom de la catégorie',
  },
  {
    label: 'Description de la catégorie',
    id: 'description',
    type: 'textarea',
    placeholder: 'Entrez la description',
  },
  {
    label: 'Image de la catégorie',
    id: 'image',
    type: 'text',
    placeholder: "URL de l'image",
  },
];

function UpdateCategoryModal({
  isOpen,
  onClose,
  category,
  onCategoryUpdated,
}: {
  isOpen: boolean;
  onClose: () => void;
  category: Category | null;
  onCategoryUpdated: (category: Category) => void;
}) {
  if (!isOpen || !category) return null;
  const [updateCategory, { loading, error }] = useUpdateCategoryMutation();

  const handleSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    const formData = new FormData(e.currentTarget);
    const formJSON = Object.fromEntries(formData.entries());

    try {
      const response = await updateCategory({
        variables: {
          id: category.id,
          name: formJSON.name.toString(),
          image: formJSON.image.toString(),
          description: formJSON.description.toString(),
        },
      });

      if (response.data?.updateCategory?.id) {
        alert('Catégorie mise à jour avec succès');
        onCategoryUpdated(response.data.updateCategory); // Utilisez la fonction ici
        onClose();
      } else {
        alert('Erreur lors de la mise à jour de la catégorie');
      }
    } catch (error) {
      alert('Erreur réseau ou de requête lors de la mise à jour de la catégorie');
      console.error('Erreur lors de la mise à jour de la catégorie', error);
    } finally {
      client.resetStore();
    }
  };

  return (
    <div>
      <input
        type="checkbox"
        id="category_modal"
        className="modal-toggle"
        checked={isOpen}
        readOnly
      />
      <div className="modal" role="dialog">
        <div className="modal-box">
          <h3 className="text-lg font-bold">Modifier une catégorie existante</h3>
          <form className="flex flex-col gap-4 p-4 border rounded" onSubmit={handleSubmit}>
            {fields.map((field) => (
              <FormInput
                key={field.id}
                id={field.id}
                label={field.label}
                placeholder={field.placeholder}
                inputType={field.type}
                defaultValue={category[field.id as keyof Category] as string}
              />
            ))}
            <button type="submit" className="btn btn-primary" disabled={loading}>
              Mettre à jour
            </button>
          </form>
        </div>
        <label className="modal-backdrop" htmlFor="category_modal" onClick={onClose}>
          Fermer
        </label>
      </div>
    </div>
  );
}

export default UpdateCategoryModal;
