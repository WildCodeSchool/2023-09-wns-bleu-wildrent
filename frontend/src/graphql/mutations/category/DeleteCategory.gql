mutation DeleteCategory($id: Int!) {
  deleteCategory(id: $id)
}
