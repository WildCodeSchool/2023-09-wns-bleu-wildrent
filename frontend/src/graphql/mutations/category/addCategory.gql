mutation AddCategory($name: String!, $description: String!, $image: String!) {
  addCategory(name: $name, description: $description, image: $image) {
    id
    name
    description
    image
  }
}
