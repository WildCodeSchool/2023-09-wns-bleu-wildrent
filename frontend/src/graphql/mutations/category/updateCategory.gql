mutation UpdateCategory($id: Int!, $name: String!, $description: String!, $image: String!) {
  updateCategory(id: $id, name: $name, description: $description, image: $image) {
    id
    name
    description
    image
  }
}
