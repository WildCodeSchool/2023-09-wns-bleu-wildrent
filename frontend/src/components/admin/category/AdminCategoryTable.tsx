import React, { useState } from 'react';
import { useMutation } from '@apollo/client';
import gql from 'graphql-tag';
import { Category } from '@/types';
import Image from 'next/image';
import UpdateCategoryModal from '@/components/admin/category/UpdateCategoryModal';
import AddCategoryModal from '@/components/admin/category/AddCategoryModal';
import client from '@/graphql/client';
import { IoIosAdd } from 'react-icons/io';
import { useAlert } from '@/components/hooks/AlertContext';

// DÃ©finition de la mutation GraphQL pour la suppression
const DELETE_CATEGORY_MUTATION = gql`
  mutation DeleteCategory($id: Int!) {
    deleteCategory(id: $id)
  }
`;

type AdminCategoryTableProps = {
  initialCategories: Category[];
};

const AdminCategoryTable: React.FC<AdminCategoryTableProps> = ({ initialCategories }) => {
  const [categories, setCategories] = useState<Category[]>(initialCategories);
  const [isModalOpen, setIsModalOpen] = useState(false);
  const [selectedCategory, setSelectedCategory] = useState<Category | null>(null);
  const [isAddModalOpen, setIsAddModalOpen] = useState(false);
  const { showAlert } = useAlert();
  const [deleteCategory] = useMutation(DELETE_CATEGORY_MUTATION);

  const handleEditClick = (category: Category) => {
    setSelectedCategory(category);
    setIsModalOpen(true);
  };

  const handleCloseModal = () => {
    setIsModalOpen(false);
    setIsAddModalOpen(false);
    setSelectedCategory(null);
  };

  const handleAddCategoryClick = () => {
    setIsAddModalOpen(true);
  };

  const handleCategoryAdded = (newCategory: Category) => {
    setCategories([...categories, newCategory]);
    handleCloseModal();
  };

  const handleCategoryUpdated = (updatedCategory: Category) => {
    const updatedCategories = categories.map((category) =>
      category.id === updatedCategory.id ? updatedCategory : category,
    );
    setCategories(updatedCategories);
    handleCloseModal();
  };

  const handleDeleteCategoryClick = async (categoryId: number) => {
    try {
      const { data } = await deleteCategory({
        variables: { id: categoryId },
      });
      if (data.deleteCategory) {
        setCategories(categories.filter((category) => category.id !== categoryId));
        showAlert('success', 'Category successfully deleted', 3000);

        client.resetStore();
      } else {
        showAlert('error', 'Failed to delete category', 3000);
      }
    } catch (error) {
      console.error('Error deleting category', error);
      showAlert('error', 'Error deleting category', 3000);
    }
  };

  return (
    <>
      <button className="btn btn-circle btn-accent" onClick={handleAddCategoryClick}>
        <IoIosAdd size={50} />
      </button>

      {isAddModalOpen && (
        <AddCategoryModal
          isOpen={isAddModalOpen}
          onClose={handleCloseModal}
          onCategoryAdded={handleCategoryAdded}
        />
      )}
      {selectedCategory && (
        <UpdateCategoryModal
          isOpen={isModalOpen}
          onClose={handleCloseModal}
          category={selectedCategory}
          onCategoryUpdated={handleCategoryUpdated}
        />
      )}
      <table className="min-w-full table-auto">
        <thead>
          <tr className="bg-gray-400 text-left text-white">
            <th className="px-4 py-2">ID</th>
            <th className="px-4 py-2">Image</th>
            <th className="px-4 py-2">Nom</th>
            <th className="px-4 py-2">Description</th>
            <th className="px-4 py-2">Actions</th>
          </tr>
        </thead>
        <tbody>
          {categories.map((category) => (
            <tr key={category.id} className={category.id % 2 === 0 ? 'bg-gray-200' : ''}>
              <td className="px-4 py-2 border-b">{category.id}</td>
              <td className="px-4 py-2 border-b">
                <Image src={category.image} width={50} height={30} alt={category.name} />
              </td>
              <td className="px-4 py-2 border-b">{category.name}</td>
              <td className="px-4 py-2 border-b">{category.description}</td>
              <td className="px-4 py-2 border-b">
                <button
                  className="mr-2 bg-blue-500 hover:bg-blue-700 text-white font-bold py-1 px-2 rounded"
                  onClick={() => handleEditClick(category)}
                >
                  Edit
                </button>
                <button
                  className="bg-red-500 hover:bg-red-700 text-white font-bold py-1 px-2 rounded"
                  onClick={() => handleDeleteCategoryClick(category.id)}
                >
                  Delete
                </button>
              </td>
            </tr>
          ))}
        </tbody>
      </table>
    </>
  );
};

export default AdminCategoryTable;
