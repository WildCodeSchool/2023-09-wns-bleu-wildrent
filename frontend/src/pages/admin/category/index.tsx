import React, { useState } from 'react';
import LayoutDashboard from '../../../components/admin/LayoutDashboard';
import AdminCategoryTable from '../../../components/admin/category/AdminCategoryTable';
import { useAllCategoriesAdminQuery } from '../../../graphql/generated/schema';
import { IoIosAdd } from 'react-icons/io';
import AddCategoryModal from '@/components/admin/category/AddCategoryModal';
import Loader from '@/components/Loader';
import { useAlert } from '@/components/hooks/AlertContext';

const CategoriesAdmin = () => {
  const { data, loading, error } = useAllCategoriesAdminQuery();
  const { showAlert } = useAlert();
  const [isModalOpen, setIsModalOpen] = useState(false);
  const handleModal = () => {
    setIsModalOpen(true);
  };

  const closeModal = () => {
    setIsModalOpen(false);
  };

  if (loading) return <Loader />;
  if (error) return showAlert('error', error?.message, 3000);

  return (
    <LayoutDashboard>
      <div className="text-center p-4">
        <h2>Categories Management</h2>
      </div>
      {data?.allCategories && <AdminCategoryTable initialCategories={data.allCategories as any} />}
    </LayoutDashboard>
  );
};

export default CategoriesAdmin;
