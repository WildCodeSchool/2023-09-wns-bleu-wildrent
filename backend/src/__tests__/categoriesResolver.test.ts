import { execute } from '../../jest.setup';
import { Category } from '../entities/category.entity';
// import getAdminContext from './helpers/getAdminContext';
import getCategories from './operations/getCategories';
import getCategoryById from './operations/getCategoryById';

describe('Categories Resolver', () => {
  it('can get a list of categories', async () => {
    await Category.create({
      id: 1,
      name: 'Mobilier',
      description:
        'Oat cake biscuit cookie chocolate pudding. Fruitcake cake cupcake bonbon pudding tiramisu jelly beans marzipan. Shortbread dessert chocolate biscuit jelly-o brownie. Toffee tootsie roll lollipop jelly beans powder croissant pie. Sesame snaps jelly-o dessert sugar plum donut caramels chocolate. Bear claw macaroon brownie cookie icing jelly sugar plum.',
      image: 'https://www.highgest.fr/image/562535501368/1920/928/Les-petits-+.jpg',
    }).save();
    await Category.create({
      id: 2,
      name: 'Eclairage',
      description:
        'Oat cake biscuit cookie chocolate pudding. Fruitcake cake cupcake bonbon pudding tiramisu jelly beans marzipan. Shortbread dessert chocolate biscuit jelly-o brownie. Toffee tootsie roll lollipop jelly beans powder croissant pie. Sesame snaps jelly-o dessert sugar plum donut caramels chocolate. Bear claw macaroon brownie cookie icing jelly sugar plum.',
      image: 'https://www.highgest.fr/image/505818799390/1920/945/Les-petits-+.jpg',
    }).save();
    const res = await execute(getCategories);
    expect(res).toMatchSnapshot();
  });

  it('can get one category by id', async () => {
    const cat3 = await Category.create({
      name: 'Sonorisation',
      description: 'Description de la catÃ©gorie Sonorisation',
      image: 'https://exemple.com/image_sonorisation.jpg',
    }).save();
    const res = await execute(getCategoryById, { categoryId: cat3.id });
    expect(res).toMatchSnapshot();
  });

  // it('can create a category', async () => {
  //   const res = await execute(
  //     addCategory,
  //     {
  //       data: {
  //         name: 'Bougie',
  //         description:
  //           'Oat cake biscuit cookie chocolate pudding. Fruitcake cake cupcake bonbon pudding tiramisu jelly beans marzipan. Shortbread dessert chocolate biscuit jelly-o brownie. Toffee tootsie roll lollipop jelly beans powder croissant pie. Sesame snaps jelly-o dessert sugar plum donut caramels chocolate. Bear claw macaroon brownie cookie icing jelly sugar plum.',
  //         image: 'https://www.highgest.fr/image/505818799390/1920/945/Les-petits-+.jpg',
  //       },
  //     },
  //     // await getAdminContext()
  //   );
  //   expect(res).toMatchSnapshot();
  // });
});
