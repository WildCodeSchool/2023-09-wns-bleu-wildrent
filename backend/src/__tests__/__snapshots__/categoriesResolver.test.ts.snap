// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Categories Resolver can get a list of categories 1`] = `
{
  "data": {
    "allCategories": [
      {
        "description": "Oat cake biscuit cookie chocolate pudding. Fruitcake cake cupcake bonbon pudding tiramisu jelly beans marzipan. Shortbread dessert chocolate biscuit jelly-o brownie. Toffee tootsie roll lollipop jelly beans powder croissant pie. Sesame snaps jelly-o dessert sugar plum donut caramels chocolate. Bear claw macaroon brownie cookie icing jelly sugar plum.",
        "id": 1,
        "image": "https://www.highgest.fr/image/562535501368/1920/928/Les-petits-+.jpg",
        "name": "Mobilier",
      },
      {
        "description": "Oat cake biscuit cookie chocolate pudding. Fruitcake cake cupcake bonbon pudding tiramisu jelly beans marzipan. Shortbread dessert chocolate biscuit jelly-o brownie. Toffee tootsie roll lollipop jelly beans powder croissant pie. Sesame snaps jelly-o dessert sugar plum donut caramels chocolate. Bear claw macaroon brownie cookie icing jelly sugar plum.",
        "id": 2,
        "image": "https://www.highgest.fr/image/505818799390/1920/945/Les-petits-+.jpg",
        "name": "Eclairage",
      },
    ],
  },
}
`;

exports[`Categories Resolver can get one category by id 1`] = `
{
  "errors": [
    [GraphQLError: Unknown argument "categoryId" on field "Query.categoryById".],
    [GraphQLError: Field "categoryById" argument "id" of type "Int!" is required, but it was not provided.],
  ],
}
`;
